### DISCUSSION TITLE
[A-L] (5 marzo 2024) Esercizio 1

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#2] [Esercitazioni A-L]
Un grafo diretto $G = (V, E)$ si dice diretto aciclico **(DAG)** se $G$ non contiene alcun ciclo diretto. Modificare l’algoritmo della ricerca
in profondità in maniera da poter controllare se un grafo diretto è aciclico o no; è possibile fare questa modifica in modo che il controllo avvenga in $\theta(|V | + |E|)$?

**Comment by LeoCosta02** [comment]
![Screenshot (4)](https://github.com/sapienzastudentsnetwork/progalg2324/assets/128373356/e3000c39-e361-48e8-9659-b2760c6f464e)

**Comment by daromacarlo** [comment]
![immagine](https://github.com/user-attachments/assets/41ba5157-49ed-4e9f-adb5-ca4725f1a240)

**Comment by FeddyLix17** [comment]
presi 2 grafi diretti di esempio

<div align="center">

![The-difference-between-a-directed-cyclic-graph-and-a-directed-acyclic-graph](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/66442720/0697d8eb-83ae-41fe-ad5f-571ea19aa98f)

</div>

implementati come

```python

graph_sx = {
              1: [2],
              2: [3],
              3: [1, 4, 5],
              4: [5],
              5: [1],
           }

graph_dx = {
              1: [2, 3, 5],
              2: [3],
              3: [4, 5],
              4: [5],
              5: [],
           }

```

(ovvero forniti delle rispettive liste di adiacenza)

l'algoritmo di verifica dell'aciclicità di un grafo può essere implementato come segue

```python

def dfs_modified(G, x, nodi_visitati, stack):
    
    # visito il nodo corrente
    nodi_visitati.append(x)
    stack.append(x)
    
    # per ogni suo nodo di uscita
    for z in G[x]:
    
        # se z non è ancora stato visitato
        if z not in nodi_visitati:
        
            # proseguo l'esplorazione
            if dfs_modified(G, z, nodi_visitati, stack):
                return True

        # altrimenti verifico la visita di z sia terminata
        # (qualora non lo fosse, ci si ritroverebbe davanti a un ciclo)
        elif z in stack:
            return True
    
    # completo la visita di x (ciclo non ancora rilevato)
    stack.pop()
    return False

def dfs_check_acyclic(G):
    
    nodi_visitati = []
    stack = []
    
    # effettuo un DFS per ogni nodo non ancora visitato
    for x in G.keys():
        if x not in nodi_visitati:
            if dfs_modified(G, x, nodi_visitati, stack):
                return False
    
    return True

print(dfs_check_acyclic(graph_sx))  # falso perchè cicliclo
print(dfs_check_acyclic(graph_dx))  # vero perchè aciclico   
 ```

non avendo stravolto più di tanto l'algoritmo [originale](https://www.geeksforgeeks.org/time-and-space-complexity-of-depth-first-search-dfs/), viene mantenuto il costo computazionale di $\Large \theta (|V|+|E|)$

**Comment by niccolo-fato** [comment]
![photo_2024-07-16_22-10-11](https://github.com/user-attachments/assets/0e064b52-49a6-4d54-90d1-75601d7ef89c)

**Comment by daromacarlo** [comment]
Un grafo diretto si dice diretto aciclico (DAG) se non contiene alcun ciclo diretto. Modificare l’algoritmo della ricerca
in profondità in maniera da poter controllare se un grafo diretto è aciclico o no; è possibile fare questa modifica in modo che il controllo avvenga in O(n+m)?

```c
def cerca_cicli(G, P = [0,...,0], v = random(V(G)), x = v){  //aux
	P[v] = -x
	for w in v.adiacenti(){
		if P[w] == 0{
			z = cerca_cicli(G, P, w ,v)
			if z != 0{
				P[v] = - P[v]
				return z
			}
		}
		else{
			if w != x{
				P[w] = 0
				P[v] = x
				return v
			}
		}
	}
	P[v] = x
	return 0
}

// costo = O(n + m)

def aciclico_test(G){
	P = [0,...,0]
	cerca_cicli(G, P, v = random(V(G)), x = v)
	i = 0
	while P[i] != 0 && i != n{
		i ++
	}
	if i == n && P[i] != 0{
		return true
	}
	else{
		return false
	}
}

```
